homeassistant/components/bryant_evolution/coordinator.py

```diff
@@ -0,0 +1,190 @@
+        """Return the current HVAC mode."""
+        evolution_mode = self.data.systems[sys_id].hvac_mode
+        mode_enum = {
+            "HEAT": HVACMode.HEAT,
+            "COOL": HVACMode.COOL,
+            "AUTO": HVACMode.HEAT_COOL,
+            "OFF": HVACMode.OFF,
+        }.get(evolution_mode.upper())
+        if mode_enum is None:
+            raise HomeAssistantError(
+                translation_domain=DOMAIN,
+                translation_key="failed_to_parse_hvac_mode",
+                translation_placeholders={"mode": evolution_mode},
+            )
+        return mode_enum
+
+    def read_target_temperatures(
+        self, sys_id: int, zone_id: int
+    ) -> tuple[int | None, int | None, int | None]:
+        """Return the target temperatures.
+
+        Specifically, returns a tuple with three integers. It will be either:
+        1. (target-temp, None, None) if the mode is HEAT or COOL
+        2. (None, heating-setpoint, cooling-setpoint) if the mode is HEAT_COOL.
+        """
+        mode = self.read_hvac_mode(sys_id)
+        # Return target_temperature or target_temperature_{high, low} based on mode.
+        match mode:
+            case HVACMode.HEAT:
+                return (self.data.systems[sys_id].zones[zone_id].htsp, None, None)
+            case HVACMode.COOL:
+                return (self.data.systems[sys_id].zones[zone_id].clsp, None, None)
+            case HVACMode.HEAT_COOL:
+                htsp = self.data.systems[sys_id].zones[zone_id].htsp
+                clsp = self.data.systems[sys_id].zones[zone_id].clsp
+                return (None, htsp, clsp)
+            case HVACMode.OFF:
+                return (None, None, None)
+            case _:
+                raise HomeAssistantError(f"Unknown HVAC mode {mode}")
+
+    def read_hvac_action(self, sys_id: int, zone_id: int) -> HVACAction:
+        """Return the HVAC action."""
+        if not self.data.systems[sys_id].is_hvac_active:
+            return HVACAction.OFF
+
+        mode = self.read_hvac_mode(sys_id)
+        match mode:
+            case HVACMode.HEAT:
+                return HVACAction.HEATING
+            case HVACMode.COOL:
+                return HVACAction.COOLING
+            case HVACMode.OFF:
+                return HVACAction.OFF
+            case HVACMode.HEAT_COOL:
+                (_, clsp, _) = self.read_target_temperatures(sys_id, zone_id)
+                assert clsp is not None
+                # In HEAT_COOL, we need to figure out what the actual action is
+                # based on the setpoints.
+                if self.read_current_temperature(sys_id, zone_id) > clsp:
+                    # If the system is on and the current temperature is
+                    # higher than the point at which heating would activate,
+                    # then we must be cooling.
+                    return HVACAction.COOLING
+                return HVACAction.HEATING
+        raise HomeAssistantError(f"Unknown mode: {mode}")
```

Comment:
This could also be a helper function instead of a method of the coordinator

```suggestion

def hvac_action(zone, mode: HVACMode | None) -> HVACAction | None:
    """Return the HVAC action."""

    mode_to_action = {
        HVACMode.HEAT: HVACAction.HEATING,
        HVACMode.COOL: HVACAction.COOLING,
        HVACMode.OFF: HVACAction.OFF,
    }

    if mode in mode_to_action:
        return mode_to_action[mode]

    if mode is HVACMode.HEAT_COOL:
        return HVACAction.COOLING if zone.curr_temp > zone.htsp else HVACAction.HEATING

```

`self._attr_hvac_action = hvac_action(self.zone, mode) if self.system.is_hvac_active else HVACAction.OFF`